File: backend/meta/datasource.py
"""
数据源实体
Datasource可以独立存在，可关联Model和Domain（可选）
"""
from typing import Optional
from datetime import date


class Datasource:
    """数据源实体 - 可独立存在"""
    
    def __init__(
        self,
        id: int,
        name: str,
        type: str,
        url: str,
        domainId: Optional[int] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
        tableName: Optional[str] = None,
        status: str = "inactive",
        description: Optional[str] = None,
        modelId: Optional[int] = None,
        createdAt: Optional[str] = None,
        updatedAt: Optional[str] = None
    ):
        self.id = id
        self.name = name
        self.type = type  # mysql, postgresql, sqlserver, duckdb
        self.url = url
        self.domainId = domainId  # 可选分类
        self.username = username
        self.password = password
        self.tableName = tableName
        self.status = status  # active, inactive
        self.description = description
        self.modelId = modelId  # 可选关联
        self.createdAt = createdAt or date.today().isoformat()
        self.updatedAt = updatedAt or date.today().isoformat()
    
    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "id": self.id,
            "name": self.name,
            "type": self.type,
            "url": self.url,
            "username": self.username,
            "password": self.password,
            "tableName": self.tableName,
            "status": self.status,
            "description": self.description,
            "modelId": self.modelId,
            "domainId": self.domainId,
            "createdAt": self.createdAt,
            "updatedAt": self.updatedAt
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Datasource':
        """从字典创建实例"""
        return cls(
            id=data["id"],
            name=data["name"],
            type=data["type"],
            url=data["url"],
            domainId=data.get("domainId"),
            username=data.get("username"),
            password=data.get("password"),
            tableName=data.get("tableName"),
            status=data.get("status", "inactive"),
            description=data.get("description"),
            modelId=data.get("modelId"),
            createdAt=data.get("createdAt"),
            updatedAt=data.get("updatedAt")
        )
    
    def toggle_status(self) -> 'Datasource':
        """切换状态"""
        self.status = "active" if self.status == "inactive" else "inactive"
        self.updatedAt = date.today().isoformat()
        return self
    
    def __repr__(self):
        return f"Datasource(id={self.id}, name='{self.name}', type='{self.type}', status='{self.status}')"

File: backend_java/src/main/java/com/jianmo/platform/entity/Datasource.java
package com.jianmo.platform.entity;

import com.jianmo.platform.common.BaseEntity;
import javax.persistence.*;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.ToString;

import java.util.ArrayList;
import java.util.List;


@Entity
@Table(name = "datasources")
@EqualsAndHashCode(callSuper = true)
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Datasource extends BaseEntity {

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String type;

    @Column(nullable = false)
    private String url;

    @Column(name = "domain_id")
    private Long domainId;

    private String username;

    private String password;

    @Column(name = "table_name")
    private String tableName;

    private String status = "inactive";

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "model_id")
    private Long modelId;

    @OneToMany(mappedBy = "datasource", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @ToString.Exclude
    private List<Mapping> mappings = new ArrayList<>();

    @OneToMany(mappedBy = "datasource", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @ToString.Exclude
    private List<ModelTableAssociation> modelTableAssociations = new ArrayList<>();

    public Datasource(String name, String type, String url, String username, String password) {
        this.name = name;
        this.type = type;
        this.url = url;
        this.username = username;
        this.password = password;
    }

    public void addMapping(Mapping mapping) {
        mappings.add(mapping);
        mapping.setDatasourceId(this.getId());
    }

    public void addModelTableAssociation(ModelTableAssociation association) {
        modelTableAssociations.add(association);
        association.setDatasourceId(this.getId());
    }

    public boolean toggleStatus() {
        this.status = "active".equals(this.status) ? "inactive" : "active";
        return "active".equals(this.status);
    }

    public boolean validateConnection() {
        if (url == null || url.isEmpty()) {
            return false;
        }
        String validTypes = "mysql,postgresql,sqlserver,duckdb";
        return validTypes.contains(type.toLowerCase());
    }
}
